### ISSUE TITLE
[QUESTION] Inconsistent Log Prob Calculation in MCore When Using Chunked vs Padded Inputs

### ISSUE THREAD

**Issue by xuhang-2** [NVIDIA/Megatron-LM#1809]
### Summary
 We are using partial rollout for LLM + RL. When computing `old_logprob` for the first chunk (`chunk1`) using MCore, we observe inconsistent results:

- `chunk1 + padding` (padded to full context length), or  
- `chunk1 + chunk2` (full sequence without padding).

Using the same model (Qwen3-1.7B, dense), same weights, and same position IDs for `chunk1`, the computed log probabilities for `chunk1` differ significantly — especially on low-probability tokens and even starting from the first token of the response.

This issue **does not occur** in Transformer native forward passes - chunked and padded runs produce identical logits for `chunk1`.

---

#### 1. MCore Self-Consistency (Single Chunk)

- When generating in **one chunk**, recomputing log probs twice yields identical results (with `temperature=1` and `0.6`) — so MCore is deterministic in isolation.

#### 2. Chunk + Padding vs Chunk + Chunk Inconsistency (Main Issue)

All experiments use **Qwen3-1.7B (dense)**.

- **Exp 1**: 
  - `output_weight`, `rotary_pos_emb` (first 10360), and `decoder_input` (first 2590 tokens) are identical.
  - But **hidden states diverge** in the first half — resembling precision/numerical instability, not logical error.

in: megatron/core/models/gpt/gpt_model.py
<img width="1877" height="618" alt="Image" src="https://github.com/user-attachments/assets/44a32450-7e64-4e1f-b58b-e8991bfa202a" />

---

### Expected Behavior

Log probabilities for `chunk1` should be **identical** regardless of whether it is:

- Run standalone + padding, or
- Run as prefix to `chunk2`.

This is fundamental for RLHF/RLAIF training where `old_logprob` must remain consistent across rollouts and re-computations.

**Comment by francistotle** [comment]
Did you get any resolution to this?
